lls/arp.c

```diff
@@ -118,21 +123,49 @@ xmit_arp_req(struct gatekeeper_if *iface, const uint8_t *ip_be,
 
 int
 process_arp(struct lls_config *lls_conf, struct gatekeeper_if *iface,
-	uint16_t tx_queue, struct rte_mbuf *buf, struct ether_hdr *eth_hdr)
+	uint16_t tx_queue, struct rte_mbuf *buf, struct ether_hdr *eth_hdr,
+	struct arp_hdr *arp_hdr)
 {
 	struct lls_mod_req mod_req;
-	struct arp_hdr *arp_hdr;
 	uint16_t pkt_len = rte_pktmbuf_data_len(buf);
+	/* pkt_in_skip_l2() already called by LLS. */
+	size_t l2_len = pkt_in_l2_hdr_len(buf);
 
-	if (pkt_len < sizeof(*eth_hdr) + sizeof(*arp_hdr)) {
+	if (pkt_len < l2_len + sizeof(*arp_hdr)) {
 		RTE_LOG(ERR, GATEKEEPER, "lls: %s interface received ARP packet of size %hu bytes, but it should be at least %zu bytes\n",
 			iface->name, pkt_len,
-			sizeof(*eth_hdr) + sizeof(*arp_hdr));
+			l2_len + sizeof(*arp_hdr));
 		return -1;
 	}
 
-	arp_hdr = rte_pktmbuf_mtod_offset(buf, struct arp_hdr *,
-		sizeof(struct ether_hdr));
+	if (iface->vlan_insert) {
+		struct vlan_hdr *vlan_hdr;
```

Comment:
1. This `if` should have an `else` clause to make sure that packets that are expected to not have a VLAN ID do not have one.

2. Since this `if` statement has grown quite a bit and it's reused almost verbatim in the ND case (i.e. `process_nd_neigh_solicitation()`), turn it into a function.